### ISSUE TITLE
Transactionally deadlock still there on 3.2.0-M1?

### ISSUE THREAD

**Issue by curreli** [slick/slick#1614]
I've been facing the #1274 issue on slick `3.1.1` so I upgraded to `3.2.0-M1` which supposedly contains a fix. However, after the upgrade, I'm still having deadlocks when using `transactionally`.

I can reproduce this issue by running more than 20 `DBIOAction` in parallel, with `transactionally`. If I remove `transactionally` then everything is fine. After investigating, I figure that 20 is the default number of threads in the `AsynExecutor` so that kind of makes sense.

``` scala
def default(name: String = "AsyncExecutor.default"): AsyncExecutor = apply(name, 20, 1000)
```

I'm using `HikariCP v2.5.1` along with `PostgreSQL Driver v9.4.1208.jre7`. PostgreSQL is configured to accept max 100 connections. The HikariCP data source is configured with a max pool size of 100.

A thread dump shows all `AsyncExecutor` threads as `RUNNABLE`:

```
"AsyncExecutor.default-20" #32 daemon prio=5 os_prio=0 tid=0x0000000024665000 nid=0x2648 runnable [0x0000000026a3e000]
   java.lang.Thread.State: RUNNABLE
        at java.net.SocketInputStream.socketRead0(Native Method)
        at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
        at java.net.SocketInputStream.read(SocketInputStream.java:170)
        at java.net.SocketInputStream.read(SocketInputStream.java:141)
        at org.postgresql.core.VisibleBufferedInputStream.readMore(VisibleBufferedInputStream.java:143)
        at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:112)
        at org.postgresql.core.VisibleBufferedInputStream.read(VisibleBufferedInputStream.java:70)
        at org.postgresql.core.PGStream.ReceiveChar(PGStream.java:283)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1799)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:200)
        - locked <0x00000006c17b10b0> (a org.postgresql.core.v3.QueryExecutorImpl)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:424)
        at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:161)
        at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:133)
        at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
        at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
        at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$InsertOrUpdateAction$$anonfun$nativeUpsert$1.apply(JdbcActionComponent.scala:559)
        at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$InsertOrUpdateAction$$anonfun$nativeUpsert$1.apply(JdbcActionComponent.scala:556)
        at slick.jdbc.JdbcBackend$SessionDef$class.withPreparedStatement(JdbcBackend.scala:352)
        at slick.jdbc.JdbcBackend$BaseSession.withPreparedStatement(JdbcBackend.scala:414)
        at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl.preparedInsert(JdbcActionComponent.scala:497)
        at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$InsertOrUpdateAction.nativeUpsert(JdbcActionComponent.scala:556)
        at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$InsertOrUpdateAction.f$1(JdbcActionComponent.scala:539)
        at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$InsertOrUpdateAction.run(JdbcActionComponent.scala:544)
        at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:31)
        at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:28)
        at slick.dbio.DBIOAction$$anon$4$$anonfun$run$3.apply(DBIOAction.scala:234)
        at slick.dbio.DBIOAction$$anon$4$$anonfun$run$3.apply(DBIOAction.scala:234)
        at scala.collection.Iterator$class.foreach(Iterator.scala:893)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
        at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
        at slick.dbio.DBIOAction$$anon$4.run(DBIOAction.scala:234)
        at slick.dbio.DBIOAction$$anon$4.run(DBIOAction.scala:232)
        at slick.basic.BasicBackend$DatabaseDef$$anon$2.liftedTree1$1(BasicBackend.scala:240)
        at slick.basic.BasicBackend$DatabaseDef$$anon$2.run(BasicBackend.scala:240)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

Before we figure this one out, is there a way I can increase the default number of threads in the `AsyncExecutor`?

**Comment by szeiger** [comment]
There is no way to configure the default AsyncExecutor. You'll have to create a separate one.

**Comment by szeiger** [comment]
This really needs a reproduction. According to your description it is https://github.com/slick/slick/blob/master/slick-testkit/src/test/scala/slick/test/jdbc/hikaricp/SlickDeadlockTest.scala#L48-L58, which was successfully fixed by #1274.

**Comment by szeiger** [comment]
I've seen this in a local test run:

```
[info] Test com.typesafe.slick.testkit.tests.ActionTest.testStreaming[h2rownum] started
Exception in thread "AsyncExecutor.default-2" java.lang.IllegalArgumentException: requirement failed: count cannot be decreased
         at scala.Predef$.require(Predef.scala:224)
         at slick.util.ManagedArrayBlockingQueue$$anonfun$decreaseInUseCount$1.apply$mcV$sp(ManagedArrayBlockingQueue.scala:53)
         at slick.util.ManagedArrayBlockingQueue$$anonfun$decreaseInUseCount$1.apply(ManagedArrayBlockingQueue.scala:52)
         at slick.util.ManagedArrayBlockingQueue$$anonfun$decreaseInUseCount$1.apply(ManagedArrayBlockingQueue.scala:52)
         at slick.util.ManagedArrayBlockingQueue.slick$util$ManagedArrayBlockingQueue$$locked(ManagedArrayBlockingQueue.scala:217)
         at slick.util.ManagedArrayBlockingQueue.decreaseInUseCount(ManagedArrayBlockingQueue.scala:52)
         at slick.util.AsyncExecutor$$anon$2$$anon$1.afterExecute(AsyncExecutor.scala:81)
         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1150)
         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
         at java.lang.Thread.run(Thread.java:745)
```

The test succeeded and it didn't cause any negative effects. This could just be an illegal state during a shutdown or something similar. But it could also point to a bug in #1274. If the in-use count is lower than it should be at some point during normal operation, this would defeat the deadlock-prevention.

**Comment by szeiger** [comment]
To make my previous comment sound less alarmist: The fact that the test succeeded and that I saw this message in the console at all (which only happens for log messages *between* tests but not the actual test output) points to an illegal shutdown state.

**Comment by hvesalai** [comment]
I've just upgraded to 3.2.0-M2 and am seeing deadlocks. I have a hikari pool with 3 connections, a queue size 1000 and 6 threads.

The form in which the deadlocks appear is

```
Future.traverse(bigCollection) { item =>
  val f1: Future = calculateSomething(item) // does not use the db, so should be unrelated

  f1.flatMap { x =>
    db.run {
      val a1: DBIOAction = compiledQueryFrom(x).result

      a1.flatMap { r =>
        val a2: DBIOAction = DBIO.sequence(manyUpdateActionsFrom(r))
        val a3: DBIOAction = insertAction

        (a1 andThen a2).transactionally
      }
    }
  }
}
```

So in essence there is a large amount of DBIOActions created at once, but not so much as to overflow the db queue.

**Comment by szeiger** [comment]
@hvesalai Are the pools configured by Slick or manually? The new scheduler needs to know the correct connection pool size. (I'm not sure we ever considered the artificial case where the number of connections is lower than the number of threads, so it's possible that it wouldn't work properly in this case.)

The gist of the code above, as far as DBIO scheduling is concerned, seems to be simply synchronous action -> wait for Future -> other synchronous action (`a1.flatMap { _ => a2 }`). Everything else should be fused.

**Comment by hvesalai** [comment]
It is a manually configured AsyncExecutor for which I hadn't configured the maxConnections. Now I changed it to have a maxConnection, but I have also made other large changes in the code to lessen the number of simultaneous DBIOActions.

I agree that everything should have been fused and so the deadlock really baffles me.

Please also see my comment in https://github.com/slick/slick/pull/1461#issuecomment-274733761

**Comment by szeiger** [comment]
So does it work now with the correctly configured size?

The `flatMap` cannot be fused. Otherwise you shouldn't run into the deadlock situation.

**Comment by hvesalai** [comment]
I have not seen any deadlocks since I configured it with maxConnections, but I also changed the code so that it does not fire simultaneous DBIOActions.

**Comment by szeiger** [comment]
From the reports so far there is no indication that the current scheduling algorithm fails if the pool is configured correctly.

I think it would be good to reimplement scheduling in 3.3 as described by @hvesalai in https://github.com/slick/slick/pull/1461#issuecomment-274733761 and myself in the subsequent comment. This should be more robust and does not require knowledge of the pool size.

**Comment by dspasojevic** [comment]
We have seen this issue when submitting DBIOs composed of large numbers of sequenced DBIOs that affect the same set of rows in the database.

We see the three worker threads attempting to read results (the same stacktrace as @curreli), but another transaction (that was initiated first) is a postgres lock point for them (i.e. the results that the three worker threads are attempting to read will not be available until the fourth transaction clears).

In the thread dump we see no thread that is attempting to read the (available) results from the transaction that is the lock point.

postgres reports the lock point transaction as "idle in transaction" - presumably because whatever should be reading the result has left.

Running the same DBIO in isolation works, and increasing the numThreads means that we can increase the number of concurrent DBIOs that can be run, but we can always hit a point where the lock point transaction comes into existence and causes the other transaction to be dead locked.

We are running 3.2.0. `numThreads` is 3 and `maxConnections` is 15.

**Comment by szeiger** [comment]
@dspasojevic Thanks, this is really useful. Can you reproduce it when `numThreads` and `maxConnections` are set to the same value? 

I hadn't really thought about this case before. This could be a deadlock in the database instead of one in Slick's scheduler. It's different from #1274 but nevertheless caused by Slick's scheduler. If my assumption is correct, it should work when `maxConnections` is not larger than `numThreads`.

Since we do our own scheduling you can run into a situation where all DBIO threads are blocked on reading from the database but the row is locked and the transaction that holds the lock is currently suspended and can't run because no DBIO thread is available. The database can't detect this as a deadlock because it assumes that all connections run independently.

I can't think of any solution that preserves our more resource-efficient scheduling with suspended connections in this case. Pinned sessions are be OK as long as they are in auto-commit mode, but when a session is in transaction mode we can't give up the DBIO thread and schedule a continuation while waiting for an asynchronous result. Instead we need to block the DBIO thread to prevent another DBIO action from running which could block until the transaction is committed.

**Comment by dspasojevic** [comment]
@szeiger I can't reproduce by setting `numThreads` and `maxConnections` to be equal.

What effect does this have on long running streaming DBIOs? I thought that the number of additional connections over `numThreads` were used as housekeeping threads to make sure streaming queries are kept alive?

Worth nothing that this issue is exacerbated by hikari thinking that it is the job of application code to detect and shutdown "idle in transaction" operations. Since neither slick or hikari do this, we can only detect and recover from this at the database level itself.

**Comment by szeiger** [comment]
Streaming doesn't usually run in a transaction so it would still free up the thread. Or is it possible that rows get locked merely by the presence of an open result set? If that's the case then we'd have to always block the thread (at least by default).

Currently all interruptions of database I/O (which are always asynchronous) give up the thread and schedule a continuation. This includes back-pressure handling during streaming as well as non-DBIO computations (for example, whenever you call a DBIO method like `map` or `flatMap` that takes an implicit `ExecutionContext`). If the session is pinned (usually because of a transaction of a streaming result set) the connection is kept alive as part of the continuation.

Setting `maxConnections` and `numThreads` to the same value means that you're basically back to the one-thread-per-connection model. Even though the scheduler can give up a thread while keeping the connection around, no other work can run in the thread pool that would prevent the continuation with the suspended connection from starting immediately because there are no more connections for that work.

So the situation where this deadlock can happen is that you have multiple DBIO actions running at the same time which contend over the same lock and have asynchronous computations while holding the lock (i.e. `a1.andThen(a2).transactionally` for two synchronous DBIO actions `a1` and `a2` is always safe but `a1.flatMap(_ => a2)` is not). The workaround is to raise `numThreads` to `maxConnections`. This creates some overhead for thread scheduling (both in the app server and the db) but it prevents the deadlock situation.

**Comment by hvesalai** [comment]
Au contraire. In postgresql streaming (the use of a cursor) requires a transaction.

**Comment by szeiger** [comment]
Ah yes, I remember needing a transaction to prevent caching of the whole result set. So I suppose without a transaction the result set gets fully materialized and there should never be a need for a lock.

**Comment by hvesalai** [comment]
Exactly so and hence if the result set is huuuuge, you want to have a transaction with the lowest applicable isolation level to prevent running out of memory.

**Comment by arobertn** [comment]
Just a data point, we are experiencing the exact issue as dspasojevic on Mar 8, and the explanation above makes perfect sense.  The underlying problem is that a thread running a statement that is waiting for a lock is blocked, due to JDBC's architecture.  No amount of cleverness by slick will make this problem go away.  Is there any way to use something like https://github.com/mauricio/postgresql-async ?

**Comment by cvogt** [comment]
@arobertn, I might be wrong, but @dspasojevic describe having a a larger connection pool than thread pool, meaning yes, connections can run out of threads and we consider that a configuration bug in the application, which we'll hopefully make harder for people to implement at some point in the future. Meaning if you increase the thread pool size for Slick's thread pool beyond the number of connections, you should be good. Am I missing something? @dspasojevic also mentioned a relationship to an "idle transaction" being involved, but without more info on the details, I don't know the impact of that.

**Comment by dspasojevic** [comment]
@cvogt  it looks like `slick.jdbc.JdbcBackend.DatabaseFactoryDef#forConfig` sets `numConnections` to be `numThreads * 5` unless overridden. This is the configuration we used when we experienced deadlocks. We set `numThreads` to `3` and allowed slick to automatically configure `numConnections` to `15`. Is using the default configuration really an application configuration bug?

"idle in transaction" is the status of a connection - as reported by postgres - that is waiting for the result to be read by slick.

When we experienced deadlocks, the "idle in transaction" connection is the one that has completed work but is still holding the database lock. That connection has been disassociated from an active slick thread and slick has scheduled a continuation to return to the connection to read the results. However, the connections that slick are actively trying to read from (e.g. are associated with one of the slick threads) are prevented from returning until the lock held by the disassociated / "idle in transaction" connection is released. This will never happen, because slick will not have an available thread to allocate to that connection.

You are correct that if the default configuration is *not* used, and I override `numThreads` and `numConnections` to be equal, then this situation can not occur, because there will always be a thread available to allocate to the connection.

**Comment by cvogt** [comment]
@dspasojevic I am probably missing something. @szeiger can you shed some light on these defaults? https://github.com/slick/slick/blob/b87abdfc4c065842cf15ec58b1c46f9ee6edae59/slick/src/main/scala/slick/jdbc/JdbcBackend.scala#L291

**Comment by arobertn** [comment]
@cvogt We were using exactly the default setup as dspasojevic: #/threads 3, conns auto conf to 15.  4 (or more) conns would get used, with 2, 3, and 4 waiting for a lock that 1 holds, but the threads are now blocked waiting for their DB ops to get the lock, and so 1 never gets run to finish.  In our case #1 (in 'idle in transaction state') has completed all DB operations and is just sending the result (from 'insert returning') over the network to the client before committing its transaction.  There is a post from Stefan on the Slick list giving some of the rationale for defaulting conns > threads: https://groups.google.com/forum/#!topic/scalaquery/Ob0R28o45eM

We will probably redesign some parts of our API to reduce the likelihood of creating this lock contention situation, but we won't be able to eliminate it completely, so we'll either go with #/threads = #/conns, or raise #/threads to lower the odds of occurrence.

**Comment by ranuser99** [comment]
We are also struggling with leaks.  It's taken a lot of our time unfortunately and still an issue.   I hope it's an application issue on our end that we're missing...

scala: 2.11.7
slick: 3.2.0
slick-hikaricp_2.11 : 3.2.0 (also tried with non slick hikari)
Oracle JVM: build 1.8.0_121-b13)
OS: Ubuntu server 16.04 up-to-date patches
DB: Postgresql 9.2.x, jdbc driver: 9.4.1212

We initialize database like this:
```
Database.forDataSource(
 datasource, 
 maxConnections, 
 AsyncExecutor(name, maxConnections * 2, maxConnections * 2, 1000, maxConnections))
```

Have also tried settings threads == maxConnections:
```
Database.forDataSource(
 datasource, 
 maxConnections, 
 AsyncExecutor(name, maxConnections, maxConnections, 1000, maxConnections))
```

Hikari datasource has these values of interest:
```
leakDetectionThreshold = 8000ms
connectionTimeout = 3000ms
validationTimeout = 2000ms
maxLifeTime = 600000ms
```

These noes are imply querying the database on a fixed interval, keeping an in-memory set of rows in sync from DB.   Some of the queries may be a few different table scans, then the results are merged into objects In memory.  We try not to map between actions and results, concerned about locking parallel connections, but it is necessary in a couple cases (one result is dependent on another).  Example:

```
    val liveRecordsAction = LiveRecordRows.to[Seq].result
    val liveRecordStreamsAction = (for {
      (liveStreamRows, liveRecordStreamRows) <- LiveStreamRows join LiveRecordStreamRows on (_.liveStreamId === _.liveStreamId)
    } yield (liveStreamRows, liveRecordStreamRows)).result
    val csIdServiceNameAction = getCSIdToServiceNameMap()
    val combinedAction = liveRecordsAction.zip(liveRecordStreamsAction.zip(csIdServiceNameAction))
    val result = db.run(combinedAction.transactionally)
    val liveRecords = result.map{
      case (liveRecordRows, (liveRecordStreamRows, csIdToServiceName)) =>
        val csIdToStreams = liveRecordStreamRows.groupBy(_._1.csId)
        val liveRecordsWithStreams = liveRecordRows.map(liveRecordRow => LiveRecord.from(liveRecordRow, csIdToStreams.getOrElse(liveRecordRow.csId, Seq.empty[(LiveStreamRow, LiveRecordStreamRow)]), csIdToServiceName(liveRecordRow.csId)))
        liveRecordsWithStreams
    }
    liveRecords
```
And another similar flow:
```
    val liveRecordsAction = LiveRecordRows.to[Seq].result
    val liveRecordStreamsAction = (for {
      (liveStreamRows, liveRecordStreamRows) <- LiveStreamRows join LiveRecordStreamRows on (_.liveStreamId === _.liveStreamId)
    } yield (liveStreamRows, liveRecordStreamRows)).result
    val csIdServiceNameAction = getCSIdToServiceNameMap()
    val combinedAction = liveRecordsAction.zip(liveRecordStreamsAction.zip(csIdServiceNameAction))
    val result = db.run(combinedAction.transactionally)
    val liveRecords = result.map{
      case (liveRecordRows, (liveRecordStreamRows, csIdToServiceName)) =>
        val csIdToStreams = liveRecordStreamRows.groupBy(_._1.csId)
        val liveRecordsWithStreams = liveRecordRows.map(liveRecordRow => LiveRecord.from(liveRecordRow, csIdToStreams.getOrElse(liveRecordRow.csId, Seq.empty[(LiveStreamRow, LiveRecordStreamRow)]), csIdToServiceName(liveRecordRow.csId)))
        liveRecordsWithStreams
    }
    liveRecords
```

Application logs, Hikari logging shows lines like this:
`Pool stats (total=6, active=0, idle=6, waiting=0)`
But over time, and some nodes are impacted more severely than others, it can quickly or slowly start to slip:
`Pool stats (total=6, active=3, idle=3, waiting=0)`
eventually:
`Pool stats (total=6, active=6, idle=0, waiting=0)`
Then exceptions will be thrown eventually.

Hikari will also report connection leaks and then return connections back to pool:
```
java.lang.Exception: Apparent connection leak detected
..
Previously reported leaked connection org.postgresql.jdbc4.Jdbc4Connection@3e322b75 was returned to the pool (unleaked)
```

The interesting thing, is that these nodes have approx. the same load and rate of DB queries.  They are simply polling the DB server on a fixed interval.  Some nodes may not have the leak exhibited or at least not for a very long time, whereas others can be exhausted in less than 24 hours.

Any thoughts?

BTW - What is the correct way to make multiple queries/updates, where each action depends on the previous result?  For example query1, then query2 depends on results of query1, then update depends on query1 and query2?  In such a scenario I have tried mapping on actions then calling db.run on the resulting composed action.

Thank you for any help!

**Comment by dspasojevic** [comment]
@cvogt do you know whether this is likely to be fixed in an upcoming release?

Is it possible for the community to provide assistance here? Happy to help, but I think that I would only be able to test possible fixes.

**Comment by cvogt** [comment]
3.2.1 just came out. Can you try it? As far as I know @szeiger and @trevorsibanda worked on related fixes.

**Comment by acjay** [comment]
I ran into this problem on my local machine, on 3.2.1. I was using `numThread = 2` and running into deadlocks with `transactionally`. I had no idea that `transactionally` could have this issue.

It should not be possible to deadlock Slick by using `transactionally`, no matter how many threads it's configured with. I don't know anything about the scheduler and those details, so maybe this isn't achievable in reality. At the very least, the semantics of scheduling should be better documented, so that it's very clear that it's possible to deadlock Slick.

**Comment by acjay** [comment]
Kind of shooting in the dark here, based on what I've read in these issues, but is it correct to say that active database connections and async Scala code scheduled as part of `DBIOAction`s both use the same thread pool? If so, maybe this is a potential design issue?

**Comment by vickumar1981** [comment]
Please feel free ignore me if this is a newbie question, but isn't hikari cp set so that the minimum # of connections is equal to numThreads (and the max, defaulted to numThreads * 5)?  If so, how can the connections drop below the number of threads?

Would adding settings into the config like:
```
hikaricp.minimumIdle = [numThreads]
hikaricp.maximumPoolSize = [numThreads]
```
fix the issue?

**Comment by hvesalai** [comment]
This issue is still there with 3.2.1. Somebody should really debug it instead of shooting in the dark. It's not like rocket science to get this kind of thing correct. The bug has now been with slick for over a year.

**Comment by nafg** [comment]
I think I'm being affected by this :(

**Comment by flojdek** [comment]
Guys, any news about this issue? I've just done some performance testing for my application and it gets stuck pretty much. I know I can always increase #numThreads but would be cool to see this fixed. Is there at least a workaround? When I set #numThreads = #maxConnections I still hit the same problem. Is it possible I'm setting #maxConnections for HikariCP, but not for Slick? I'm not sure how to override Slick's #maxConnections value (using Play Framework 2.5.12).

**Comment by hvesalai** [comment]
@szeiger we are now seeing the "count cannot be increased" error you reported earlier in production. So it is not just related to shutdown as you hoped.

```
Exception in thread "background-15" java.lang.IllegalArgumentException: requirement failed: count cannot be increased
        at scala.Predef$.require(Predef.scala:224)
        at slick.util.ManagedArrayBlockingQueue$$anonfun$increaseInUseCount$1.apply$mcV$sp(ManagedArrayBlockingQueue.scala:40)
        at slick.util.ManagedArrayBlockingQueue$$anonfun$increaseInUseCount$1.apply(ManagedArrayBlockingQueue.scala:39)
        at slick.util.ManagedArrayBlockingQueue$$anonfun$increaseInUseCount$1.apply(ManagedArrayBlockingQueue.scala:39)
        at slick.util.ManagedArrayBlockingQueue.slick$util$ManagedArrayBlockingQueue$$locked(ManagedArrayBlockingQueue.scala:217)
        at slick.util.ManagedArrayBlockingQueue.increaseInUseCount(ManagedArrayBlockingQueue.scala:39)
        at slick.util.AsyncExecutor$$anon$2$$anon$1.beforeExecute(AsyncExecutor.scala:74)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1146)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

**Comment by hvesalai** [comment]
I'm happy to say that we are no longer seeing this in our fork (https://github.com/hvesalai/slick/commits/3.2.1-TMNOW) when using `maxThreads = minThreads`. See #1807 for analysis on the effect of `minThread < maxThread`

**Comment by jontra** [comment]
@hvesalai We are also seeing this error in production. Been trying to use your patch, and it works, but only when `maxThreads == minThreads == maxConnections` (`minThreads == maxThreads` is not enough)

**Comment by hvesalai** [comment]
Ah yes, that is true. It must be that `maxThreads == maxConnections` and now that we have found that `minThreds == maxThreads` it means all three must be the same not to trigger bugs. There was discussion about `maxConnections` in this issue up here https://github.com/slick/slick/issues/1614#issuecomment-284730145

**Comment by arobertn** [comment]
Where are you configuring minThreads and maxThreads?  There is scala.concurrent.context.min/maxThreads but the Slick/Hikari config only exposes numThreads.  I assume this gets passed down to the relevant execution context internally.  Is setting that not equivalent to setting min/maxThreads to same?

**Comment by hvesalai** [comment]
min/maxThreads are defined in slick.util.AsyncExecutor

**Comment by jontra** [comment]
@arobertn There is no configurations for max/min thread, only `numThreads` (see `AsyncExecutor`, it is always called with `numThreads` both for `minThreads` and for `maxThreads`.

**Comment by jontra** [comment]
@hvesalai When we set `numThread == maxConnection` we stopped seeing this exception (even without your patch (https://github.com/hvesalai/slick/commits/3.2.1-TMNOW).
I am still not sure why. any ideas?

**Comment by hvesalai** [comment]
@jontra the fix we are using only affects AndThen and Seq. If you are not using either, then that could be one reason.

**Comment by hvesalai** [comment]
I have created a series of PRs that most probably fix this or reveal bad configurations leading to this. Please everybody take a look and if you can, test.

https://github.com/slick/slick/pull/1853
https://github.com/slick/slick/pull/1854
https://github.com/slick/slick/pull/1855

**Comment by stubbi** [comment]
Unfortunately, we are still running into this issue after upgrading to slick 3.2.2. 😞 Not running our queries as `.transactionally` solves the issue. We are not touching the default thread and connection settings.

**Comment by hvesalai** [comment]
@stubbi can you open a new issue and describe all the relevant settings (maxThread, minThread, maxConnections). Have you looked at the logs that there are no warnings about settings that could cause deadlocks?

**Comment by hvesalai** [comment]
Also note that this issue was about deadlocks in the slick thread pool. If your deadlock happens somewhere lese (e.g. in the DB), that is completely separate issue. Slick cannot prevent deadlocks that it has no control over, such as, for example, those caused by inapropriate use of DB-level locking (implicit or explicit).

**Comment by stubbi** [comment]
@hvesalai I described our problem with more details in https://github.com/slick/slick/issues/1890

**Comment by jimm-porch** [comment]
Although this is supposedly fixed in 3.3, I'm stuck on 3.2 until the [`java.time.Instant` issue 2053](https://github.com/slick/slick/issues/2053) is fixed.

[I was hopeful](https://github.com/slick/slick/issues/1614#issuecomment-364005574) that if I configured the `minThreads = maxThreads = maxConnections` that this would solve the problem but even after doing that I'm still seeing this exception with 3.2.3:

```
Exception in thread "CorePostgresDriver.AsyncExecutor-8" java.lang.IllegalArgumentException: requirement failed: count cannot be decreased
	at scala.Predef$.require(Predef.scala:224)
	at slick.util.ManagedArrayBlockingQueue$$anonfun$decreaseInUseCount$1.apply$mcV$sp(ManagedArrayBlockingQueue.scala:54)
	at slick.util.ManagedArrayBlockingQueue$$anonfun$decreaseInUseCount$1.apply(ManagedArrayBlockingQueue.scala:53)
	at slick.util.ManagedArrayBlockingQueue$$anonfun$decreaseInUseCount$1.apply(ManagedArrayBlockingQueue.scala:53)
	at slick.util.ManagedArrayBlockingQueue.locked(ManagedArrayBlockingQueue.scala:201)
	at slick.util.ManagedArrayBlockingQueue.decreaseInUseCount(ManagedArrayBlockingQueue.scala:53)
	at slick.util.AsyncExecutor$$anon$2$$anon$1.afterExecute(AsyncExecutor.scala:129)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1150)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
```

Here's how I'm configuring my `DatabaseDef`, and I'm passing `poolSize` of 10, which is the same value I provide to Hikari as the max pool size:

```scala
val asyncExecutor = AsyncExecutor(
  name = "CorePostgresDriver.AsyncExecutor",
  minThreads = poolSize,
  maxThreads = poolSize,
  maxConnections = poolSize,
  queueSize = 1000 // the default used by AsyncExecutor.default()
)
api.Database.forDataSource(dataSource, Some(poolSize), executor = asyncExecutor)
```

**Comment by jimm-porch** [comment]
By coincidence some of my coworkers encountered transactional issues with Slick today.  Instead of seeing the "count cannot be decreased" errors like I did, they saw timeouts waiting on connections from the pool.  It seemed to behave OK under light load but under regular use it quickly hangs.  In the DB itself we only see connections in state "idle in transaction", no obvious locks, as though `commit` were never invoked.

The `AsyncExecutor` was configured as I show above.

As far as they can tell, they are not aware of any DB locking.  For now they are removing `transactionally` from the code to get things working.
